### ISSUE TITLE
Deprecation Warning: The legacy JS API is deprecated and will be removed in Dart Sass 2.0.0

### ISSUE THREAD

**Issue by doberkofler** [skovy/typed-scss-modules#232]
Your issue may already be reported!
Please search on the [issue tracker](../) before creating one.

## Expected Behavior

After upgrading to sass 1.79.1, i see the following warning:
Deprecation Warning: The legacy JS API is deprecated and will be removed in Dart Sass 2.0.0.

## Your Environment

- Version used: 8.0.1
- Operating System and versions: macOS 15.0

**Comment by terrymun** [comment]
It seems like the deprecation notice is coming from the consumption of `LegacySyncImporter` and `LegacySyncImporterThis` in two instances in the source code: in [`importer.ts`](https://github.com/skovy/typed-scss-modules/blob/c60f74e51977e48fae30e8b868d417c18e27b8be/lib/sass/importer.ts#L2) and in [`importer.test.ts`](https://github.com/skovy/typed-scss-modules/blob/c60f74e51977e48fae30e8b868d417c18e27b8be/__tests__/sass/importer.test.ts#L2).

Perhaps a migration isn't as painful as it seems? (famous last words)

**Comment by am0wa** [comment]
At least provide a passthrough option to silence it out, Please.

**Comment by cryptofelon** [comment]
> At least provide a passthrough option to silence it out, Please.

PLEASE.

**Comment by cryptofelon** [comment]
Turns out the error goes away if you change your implementation from `sass` (default) to `node-sass`.

```
scssTyper("src/**/*.scss", { updateStaleOnly: true, logLevel: "silent", implementation: "node-sass" });

```

**Comment by vilindberg** [comment]
I created a patch that uses the new apis instead: https://gist.github.com/vilindberg/f620f5453c74aa856762dcc8ce0a6435

**Comment by skovy** [comment]
this is an open source project, changes can be opened as pull requests! i don't use this project, nor get paid to maintain it so there will be lag to reviewing and merging. if in a pinch forking and patching are also options!

**Comment by crgwbr** [comment]
> this is an open source project, changes can be opened as pull requests! i don't use this project, nor get paid to maintain it so there will be lag to reviewing and merging. if in a pinch forking and patching are also options!

@skovy If you're interested moving this project to a new maintainer, my team at the @thelabnyc might be interested. We do use this tool on quite a few projects, so we'd have a basis to justify the investment. Feel free to email me to discuss more: cweber at thelab dot co.

**Comment by msmx-mnakagawa** [comment]
Although I'm not sure how difficult to do it so much for now,
as a trial, could I start making a PR applying the patch suggested by @vilindberg not into `dist` directory but into `lib` directory?

**Comment by krzysztof-cislo** [comment]
@msmx-mnakagawa are you going to create a PR soon?

**Comment by msmx-mnakagawa** [comment]
@krzysztof-cislo
> are you going to create a PR soon?

Thank you for seeking confirmation.
I hadn't noticed some 👍 reactions which might indicate agreement with my post. 🥲

Therefore, I'll start trying to work on creating a PR from now.
By the end of the day on 4/14 (Mon) JST, at the latest, I'll comment with either a notification of the PR creation or an update if I find it difficult to do so.

**Comment by msmx-mnakagawa** [comment]
(cc: @krzysztof-cislo @shamsartem)

I've tried creating a PR #239 to address this issue. 🙌

While it is still draft as it may require additional consideration,
it should at least be able to provide verification of whether the PR operates in each specific environment.

If possible, could someone take a look at the PR? 👀

**Comment by msmx-mnakagawa** [comment]
Hi, @skovy.
Kindly pinging you to inform that I've made PRs related to `Node Sass`.

If it's convenient, would you mind helping to review #240 to #239?
(If not, my apologies for the disturbance. 🙇‍♂️)

**Comment by krzysztof-cislo** [comment]
> While it is still draft as it may require additional consideration, it should at least be able to provide verification of whether the PR operates in each specific environment.

I can test both PR versions locally, @msmx-mnakagawa . Can you provide me built npm packages? It could be easier than checkout your forks and build manually.

**Comment by msmx-mnakagawa** [comment]
@krzysztof-cislo
> Can you provide me built npm packages?

Sure! 🙆‍♂

I've created releases for both PRs from my repository.
- [v0.1.0-drop-node-sass-support](https://github.com/msmx-mnakagawa/typed-scss-modules/releases/tag/v0.1.0-drop-node-sass-support)
- [v0.2.1-migrate-dart-sass-to-resolve-deprecation-warning](https://github.com/msmx-mnakagawa/typed-scss-modules/releases/tag/v0.2.1-migrate-dart-sass-to-resolve-deprecation-warning)

We can install them as packages by the following commands.
If anything is unclear or problematic during the installation, please let me know.

### v0.1.0-drop-node-sass-support (#240)

```shell
npm install -D https://github.com/msmx-mnakagawa/typed-scss-modules/releases/download/v0.1.0-drop-node-sass-support/typed-scss-modules-0.1.0-drop-node-sass-support.tgz
```

### v0.2.1-migrate-dart-sass-to-resolve-deprecation-warning (#239)

```shell
npm install -D https://github.com/msmx-mnakagawa/typed-scss-modules/releases/download/v0.2.1-migrate-dart-sass-to-resolve-deprecation-warning/typed-scss-modules-0.2.1-migrate-dart-sass-to-resolve-deprecation-warning.tgz
```

<details>
<summary>Old Version for #239</summary>

### v0.2.0-migrate-dart-sass-to-resolve-deprecation-warning (#239)

We can confirm the package of this PR as well.
However, note that we seemingly need to install node-sass additionally for this package.

```shell
npm install -D https://github.com/msmx-mnakagawa/typed-scss-modules/releases/download/v0.2.0-migrate-dart-sass-to-resolve-deprecation-warning/typed-scss-modules-0.2.0-migrate-dart-sass-to-resolve-deprecation-warning.tgz
npm install -D node-sass // temporarily needed, which is going to be fixed
```

IMO, this may be due to the change for some existing logic around `getImplementation()`, which I'm going to resolve after #240 is merged.

</details>

**Comment by krzysztof-cislo** [comment]
Thanks @msmx-mnakagawa !

I have checked both PR versions.

The version `v0.1.0-drop-node-sass-support` is working fine. I mean that I have regenerated all types and nothing has changed comparing to types generated from the latest official version. I still can see deprecation warning about legacy-js-api in the console.

The versions `v0.2.0-migrate-dart-sass-to-resolve-deprecation-warning` is also working fine as above (results are identical). Also, I can't see any warning in the console.
Just a small note, to install `node-sass` we have to use Node 18.x or lower as it doesn't support 20.x or 22.x.

My environment is Windows 11 with WSL 2.

**Comment by msmx-mnakagawa** [comment]
Thank you for your quick confirmation, @krzysztof-cislo!

Regarding the additional installation of `Node Sass`, I've created a fixed version.
Could you try the following command at your convenience, if possible?

```.shell
npm install -D https://github.com/msmx-mnakagawa/typed-scss-modules/releases/download/v0.2.1-migrate-dart-sass-to-resolve-deprecation-warning/typed-scss-modules-0.2.1-migrate-dart-sass-to-resolve-deprecation-warning.tgz
```

With this version, you won't need to install `node-sass` additionally, unless you specify `node-sass` as your `implementation`.

**Comment by krzysztof-cislo** [comment]
@msmx-mnakagawa , I have checked `typed-scss-modules-0.2.1-migrate-dart-sass-to-resolve-deprecation-warning` version. I confirm that I didn't have to install `node-sass` (for `sass` implementation). Also, types were still generated correctly.

**Comment by msmx-mnakagawa** [comment]
@krzysztof-cislo
OK, thank you!

Now, let's anticipate whoever will step forward to review the PR. 👍

**Comment by msmx-mnakagawa** [comment]
@skovy
If you are currently unable to review or maintain this repository, would you be okay with me publishing a fixed version that addresses some `Node Sass`?
I'd do this as a separate npm package from my forked repository, if I can.

I've read your previous post https://github.com/skovy/typed-scss-modules/issues/232#issuecomment-2558155392, but I'd like to confirm just to be sure.

**Comment by krzysztof-cislo** [comment]
> [@skovy](https://github.com/skovy) If you are currently unable to review or maintain this repository, would you be okay with me publishing a fixed version that addresses some `Node Sass`? I'd do this as a separate npm package from my forked repository, if I can.
> 
> I've read your previous post [#232 (comment)](https://github.com/skovy/typed-scss-modules/issues/232#issuecomment-2558155392), but I'd like to confirm just to be sure.

@skovy , can you tell us if there is any chance for you to review @msmx-mnakagawa changes? A lot of effort has been put into these changes so it would be nice to merge them.

**Comment by inoyakaigor** [comment]
PR's from @msmx-mnakagawa can be published under \@next tag and so this PR's can be tested in the wild. In case of issues someone come here to fill new bug

**Comment by krzysztof-cislo** [comment]
> PR's from [@msmx-mnakagawa](https://github.com/msmx-mnakagawa) can be published under @next tag and so this PR's can be tested in the wild. In case of issues someone come here to fill new bug

Good idea @inoyakaigor !

@skovy can you do it, please?

**Comment by krzysztof-cislo** [comment]
@skovy , can you create a new release, please?

**Comment by krzysztof-cislo** [comment]
> this is an open source project, changes can be opened as pull requests! i don't use this project, nor get paid to maintain it so there will be lag to reviewing and merging. if in a pinch forking and patching are also options!

@skovy , nobody expects you to review, merge or create release within hours or event days, @msmx-mnakagawa created a PR a month ago and put lot of effort to analyze, modify and test everything. 

If you don't have time for this project simply pass it to someone who could at least react in reasonable time. @crgwbr volunteered to be a maintainer. If you don't want to let anyone continue your project, just update the readme file that this project is no longer maintained. I am sure that someone (maybe even @crgwbr ) will simply fork it under different name and continue it.

Right now we are in a situation where developers want to contribute to the project but it's for nothing. 😢

**Comment by Daniel3711997** [comment]
Any news here?

**Comment by doberkofler** [comment]
I personally feel that forking the project might be the way to go.

**Comment by krzysztof-cislo** [comment]
Any volunteer? @crgwbr , you wrote that you are interested take over the maintenance of this project?

**Comment by crgwbr** [comment]
@krzysztof-cislo Yeah, I think I'm going to fork it. Will post here again once there's progress on that front.

**Comment by crgwbr** [comment]
Ok, initial version of the fork is up here: https://www.npmjs.com/package/@thelabnyc/typed-scss-modules/v/8.2.0-a2. 

Main repo is here: https://gitlab.com/thelabnyc/typed-scss-modules, but is also mirrored to https://github.com/thelabnyc/typed-scss-modules.

**Comment by msmx-mnakagawa** [comment]
Thank you for the quick action, @crgwbr! ✨

As far as I can tell from [the merge commit](https://github.com/thelabnyc/typed-scss-modules/commit/a7de266bd3836ac167574193090c9d7aee67c5dc) in the fork, it already includes #240.
So, building on that, I'll go ahead and try creating a new PR for the fork that addresses #239 and resolves this issue.

Please bear with me.

**Comment by crgwbr** [comment]
That one should already be handled as well actually. But please test and
let me know if I missed anything.

**Comment by msmx-mnakagawa** [comment]
@crgwbr
Oh, I just overlooked your commit around https://github.com/thelabnyc/typed-scss-modules/commit/ba109a8fe14b31ae33a48e6305162159713414eb. 🙇‍♂

I've tested the fork in my repository and confirmed that it worked without deprecation warnings. 👍
Thank you for your effort!

**Comment by msmx-mnakagawa** [comment]
Hi, @crgwbr. 👋
Thank you again for forking!

Apologies for the delay, but I took a look at your fork in response to your suggestion below:

> That one should already be handled as well actually. But please test and let me know if I missed anything.

Based on that, I've created a few PRs addressing the points I noticed.
I'd appreciate it if you could confirm them when you have a moment. 🙇‍♂

Firstly, some PRs I created are related to this issue #232:
- [thelabnyc#1](https://github.com/thelabnyc/typed-scss-modules/pull/1)
- [thelabnyc#3](https://github.com/thelabnyc/typed-scss-modules/pull/3) (← [thelabnyc#2](https://github.com/thelabnyc/typed-scss-modules/pull/2))

Additionally, the remaining addresses an issue that existed before forking:
- [thelabnyc#4](https://github.com/thelabnyc/typed-scss-modules/pull/4)

<details>
<summary>Notes</summary>

- I created these PRs on your Git**Hub** mirror. In addition, I didn't create corresponding issues there because I likely don't have permission to access the issues page on that mirror.
    Please let me know if you'd prefer I contribute to Git**Lab**, or if you'd like me to create issues alongside the PRs.
- For readability of this issue's hitstory, if there are any detailed discussions about each PR, it would be better to continue them within each respective PR rather than on this page
</details>

**Comment by crgwbr** [comment]
Thanks, @msmx-mnakagawa! I'll start looking over these, but it'd be great if you could push these to gitlab instead of github. We really treat GitHub as a readonly mirror.

**Comment by terrymun** [comment]
I don't mean to detract from the thread, but while trying to get around the legacy API errors, I realized that I could use a VSCode plugin called [SCSS IntelliSense](https://marketplace.visualstudio.com/items?itemName=mrmlnc.vscode-scss) that infers CSS module types for me without needing to build the definition files.

Of course, this comes with major disadvantages such as:

1. It is IDE-specific, so if you have a mixed team of developers who each prefer different IDEs, this approach will not work and generating static type definitions is still the way to go
2. It adds overhead when starting up the IDE, which can be taxing on slower machines

I still believe this plugin will benefit a lot of users, but just offering a different escape hatch to those who might find utility in it.

**Comment by msmx-mnakagawa** [comment]
@crgwbr
> it'd be great if you could push these to gitlab

Oh, my apologies for the inconvenience! 😭

I've closed the existing PRs that were targeting GitHub.
Instead, I've recreated PRs [#13](https://gitlab.com/thelabnyc/typed-scss-modules/-/merge_requests/13), [#15](https://gitlab.com/thelabnyc/typed-scss-modules/-/merge_requests/15) (← [#14](https://gitlab.com/thelabnyc/typed-scss-modules/-/merge_requests/14)), and [#16](https://gitlab.com/thelabnyc/typed-scss-modules/-/merge_requests/16), along with their corresponding issues, directly on GitLab.

Sorry for all the back and forth, but please take a look when you get a chance. 👍
